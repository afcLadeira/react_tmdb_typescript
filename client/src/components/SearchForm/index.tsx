import React, { useEffect } from "react";
import { useForm, Controller, SubmitHandler } from "react-hook-form";

import Button from "react-bootstrap/Button";
import Form from "react-bootstrap/Form";
import { useNavigate } from "react-router-dom";
import { RESULTS_ROUTE } from "../../constants";
import SearchResults from "../SearchResults";
import useDebouncer from "../../hooks/useDebouncer";

import windowIcon from "../../assets/window.png";
import { Image, InputGroup } from "react-bootstrap";

interface SearchFormProps {
  setShowPopular: React.Dispatch<React.SetStateAction<boolean>>;
}

export type SearchFormValues = {
  multi: boolean | undefined;
  searchText: string;
};

export default function SearchForm({ setShowPopular }: SearchFormProps) {
  let navigate = useNavigate();

  const {
    control,
    reset,
    register,
    handleSubmit,
    watch,
    formState: { errors },
  } = useForm<SearchFormValues>();

  function onFormSubmit(data: SearchFormValues) {
    navigate(`${RESULTS_ROUTE}${data.searchText}`, {
      state: { multi: data.multi },
    });
  }

  // function onFormSubmit(data : SearchFormValues) {

  //     navigate(`${RESULTS_ROUTE}${data.searchText}`, {
  //       state: { multi: data.multi === "true" ? true : false },
  //     });
  //   }

  const onSubmit: SubmitHandler<SearchFormValues> = (data) =>
    onFormSubmit(data);

  const watchAllFields = watch();

  const debounceValue = useDebouncer(
    watchAllFields.searchText,
    watchAllFields.multi
  );

  useEffect(() => {
    if (watchAllFields.searchText) {
      setShowPopular(false);
    } else {
      setShowPopular(true);
    }
  }, [setShowPopular, watchAllFields.searchText]);

  return (
    <div data-testid="searchForm" style={{ padding: 20 }}>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Controller
          control={control}
          name="searchText"
          render={({ field: { onChange, onBlur, value, ref } }) => (
            <Form.Group className="mb-3" controlId="searchText">
              <InputGroup>
                <Form.Control
                  data-testid="searchInput"
                  isInvalid={!!errors.searchText}
                  as="input"
                  type="text"
                  placeholder="Search"
                  {...register("searchText", {
                    required: "This is required",
                  })}
                />
                <Button
                  onClick={() => reset()}
                  variant="outline-secondary"
                  id="button-addon2"
                >
                  Clear
                </Button>
              </InputGroup>
              <Form.Control.Feedback type="invalid">
                {errors.searchText && errors.searchText.message}
              </Form.Control.Feedback>
            </Form.Group>
          )}
        />
        <div style={{ display: "flex", justifyContent: "space-between" }}>
          <Controller
            control={control}
            name="multi"
            render={({ field: { onChange, onBlur, value = false, ref } }) => (
              <Form.Group className="mb-3" controlId="muti">
                <Form.Check
                  type="checkbox"
                  label="All media types"
                  //defaultChecked={false}
                  checked={value}
                  isInvalid={!!errors.multi}
                  {...register("multi")}
                />
              </Form.Group>
            )}
          />

          <Button variant="light" size="sm" type="submit">
            {" "}
            <Image width={20} src={windowIcon}></Image>
            {/* Search {watchAllFields.multi ? "all media" : "movie"} */}
          </Button>
        </div>
      </Form>
      {/* for jest test */}
      {debounceValue && <h1 data-testid="displaySearch">{debounceValue}</h1>}

      <SearchResults
        searchString={debounceValue}
        multi={watchAllFields.multi}
      ></SearchResults>
    </div>
  );
}
