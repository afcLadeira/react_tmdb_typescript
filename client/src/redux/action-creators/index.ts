import { AxiosInstance } from "axios";
import { Dispatch } from "redux";

export const getFavorites = (data: any) => {
  return async (dispatch: Dispatch) => {
    dispatch({
      type: "getFavorites",
      payload: data,
    });
  };
};

export const addToFavorites = (item: any, axiosPrivate: AxiosInstance) => {
  return async (dispatch: Dispatch, getState: any) => {
    await axiosPrivate.post(
      `/api/favorites/${getState().favorites.userId}`,
      item
    );

    dispatch({
      type: "addFavorite",
      payload: item,
    });
  };
};

export const removeFromFavorites = (item: any, axiosPrivate: AxiosInstance) => {
  return async (dispatch: Dispatch, getState: any) => {
    await axiosPrivate.delete(
      `/api/favorites/${getState().favorites.userId}/${item.id}`
    );

    dispatch({
      type: "removeFavorite",
      payload: item,
    });
  };
};
