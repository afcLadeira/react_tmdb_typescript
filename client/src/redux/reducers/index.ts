import { combineReducers } from 'redux'
import favoritesReducer from "./favoritesReducer"



const reducers = combineReducers({
    favorites : favoritesReducer,
})


export default reducers