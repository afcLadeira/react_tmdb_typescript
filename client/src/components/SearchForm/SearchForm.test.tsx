import {
  render,
  screen,
  cleanup,
  act,
  fireEvent,
  waitFor,
} from "@testing-library/react";

import renderer from 'react-test-renderer';

import SearchForm from ".";

import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { store } from "../../redux/store";

import { QueryClient, QueryClientProvider } from "react-query";
//import SearchResults from "../SearchResults";

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
      staleTime: Infinity,
    },
  },
});

afterEach(cleanup);

jest.setTimeout(10000);

       // Set up a Jest mock function to check any props.
// const mockChildComponent = jest.fn();

// // Mock the child component file, and grab all props passed to it
// // and pass them to the Jest mock function so it can listen.
// jest.mock("../SearchResults", () => (props) => {
//   mockChildComponent(props);

//   // Return something for React to render. I like creating
//   // an element to pass through as it
//   // helps with visually debugging if needed.
//   return <mock-childComponent />;
// });

describe("Search Form Component", () => {
  function tree() {
    return render(
      <Provider store={store}>
        <BrowserRouter>
          <QueryClientProvider client={queryClient}>
            <SearchForm
              setShowPopular={() => console.log("showform prop function")}
            ></SearchForm>
          </QueryClientProvider>
        </BrowserRouter>
      </Provider>
    );
  }

  it("rendered Form", () => {
    tree();

    const element = screen.getByTestId("searchForm");
    //expect(element).toBeTruthy();
    //or
    expect(element).toBeInTheDocument();
  });

  it("rendered search input", () => {
    tree();

    const input = screen.getByTestId("searchInput");

    expect(input).toBeInTheDocument();
  });

  it("show search string after input change", async () => {

    
    tree();

    const input = screen.getByTestId("searchInput");

    const inputWord = "matrix";

    //dont need act in this case if using waitFor
    // await act(async () => {
    fireEvent.change(input, { target: { value: inputWord } });
    // });

    //test input directly
    // expect(input.value).toBe('matrix')

    await new Promise((r) => setTimeout(r, 3000));

    await waitFor(() => {
      const searchTextResults = screen.getByTestId("displaySearch");
      expect(searchTextResults.innerHTML).toBe(inputWord);
    });

  });

  it("show results after input change", async () => {
    tree();

    const inputElement = screen.getByTestId("searchInput");

    const inputWord = "matrix2";

    fireEvent.change(inputElement, { target: { value: inputWord } });

    await new Promise((r) => setTimeout(r, 3000));

    await waitFor(() => {
        
      //expect(screen.getByTestId('childComponent')).toBeInTheDocument();

      const searchTextResults2 = screen.getByTestId(
        "displaySearchOtherComponent"
      );
      expect(searchTextResults2.innerHTML).toBe(inputWord);
    });
  });
  it("match snapshot", async () => {

    const treeSnapshot = renderer.create(<Provider store={store}>
        <BrowserRouter>
          <QueryClientProvider client={queryClient}>
            <SearchForm
              setShowPopular={() => console.log("showform prop function")}
            ></SearchForm>
          </QueryClientProvider>
        </BrowserRouter>
      </Provider>).toJSON()
    expect(treeSnapshot).toMatchSnapshot()

  })

    // it("---show results after input change", async () => {

    //   tree();

    //   const inputElement = screen.getByTestId("searchInput");

    //   const inputWord = "lost";

    //     fireEvent.change(inputElement, { target: { value: inputWord } });

    //     await new Promise((r) => setTimeout(r, 5000));

    //     await waitFor(() => {
    //     expect(mockChildComponent).toHaveBeenCalledWith(
    //       expect.objectContaining({
    //         multi: false,
    //         searchString: "lost"
    //       })
    //     );
    //   })

    // });
});
