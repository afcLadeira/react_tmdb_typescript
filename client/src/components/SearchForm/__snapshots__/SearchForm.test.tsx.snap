// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Form Component match snapshot 1`] = `
<div
  data-testid="searchForm"
  style={
    Object {
      "padding": 20,
    }
  }
>
  <form
    className=""
    onSubmit={[Function]}
  >
    <div
      className="mb-3"
    >
      <div
        className="input-group"
      >
        <input
          className="form-control"
          data-testid="searchInput"
          id="searchText"
          name="searchText"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Search"
          type="text"
        />
        <button
          className="btn btn-outline-secondary"
          disabled={false}
          id="button-addon2"
          onClick={[Function]}
          type="button"
        >
          Clear
        </button>
      </div>
      <div
        className="invalid-feedback"
      />
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "justifyContent": "space-between",
        }
      }
    >
      <div
        className="mb-3"
      >
        <div
          className="form-check"
        >
          <input
            checked={false}
            className="form-check-input"
            disabled={false}
            id="muti"
            name="multi"
            onBlur={[Function]}
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="form-check-label"
            htmlFor="muti"
            title=""
          >
            All media types
          </label>
        </div>
      </div>
      <button
        className="btn btn-light btn-sm"
        disabled={false}
        type="submit"
      >
         
        <img
          className=""
          src="window.png"
          width={20}
        />
      </button>
    </div>
  </form>
</div>
`;
